### DISCUSSION TITLE
Feature Requests

### DISCUSSION THREAD

**Discussion by amandaghassaei** [amandaghassaei/OrigamiSimulator#41] [Ideas]
Hi everybody!  Sorry I've been a bit MIA the last few years, I really appreciate the work others have done to fix bugs and keep PRs moving since I've been away!  I'm thinking that I might have some time later this year to do some work on this codebase.  I'd love to compile a list of feature requests and try to figure out how to prioritize them in this thread.  Based on emails I've received over time, here are a few that seem useful, please let me know if you agree/disagree or have other ideas!

- Ability to connect edges in a design so that you can import a tessellation as SVG and turn it into e.g. a connected tube in 3D.  Currently this type of thing is only supported through fold, but not everyone wants to write fold files.
- Separate the simulation and other mathy stuff into a core library that can be imported into other projects (via npm) and runs headless.
- Better triangulation library, or smarter triangulation algs in general that anticipate the bending behavior of developable surfaces.  Maybe a dynamic triangulation that discretizes the surface more where bending is detected.  Could also take a look at e.g. geodesic nets approach https://www.youtube.com/watch?v=yEpHoZyFhnk
- I want to move the whole codebase to Typescript/es6 with webpack and use something a bit more modern for the ui (e.g. react).
- Ability to select from multiple solvers in the app so other simulation methods could be supported, maybe we can get more buy-in from outside contributors.  I would love to get something in here similar to the rigid origami simulator that does not require the fold angles to be known.
- I don't think any of the web-vr stuff is supported anymore, so let's just rip that all out for now.  I think we may want to move vr stuff into another app entirely.
- Import texture maps to add designs to the surface of the forms.
- Support more options for rendering out images / animations.  e.g. control of lighting, shadows, materials, etc.
- Fold angle sequences – support FOLD files with frames to control a sequence of fold angles.  We need to figure out how to support fold sequences through SVG import (it might make sense to build some simple keyframing editor to specify the folding sequence through the app).
- Exporting keyframed 3d models (gltf) for use in other apps.
- Collision detection?
- Exporting vector images of 3D model for use in paper figures and other diagrams.
- Hover over edges/faces and see their index in FOLD, assignment, current dihedral, target dihedral, etc - would be useful for debugging bad files.  This can def be in "advanced" mode.

questions I have (so far):

- If we were to enable some kind of design editing interface - what types of edits would people want to make? 

**Edit (Dec 2021):** Well there goes 2021 and I was not able to make time for this.  Unfortunately the beginning half of next year is a bit busy, but planning to get back to this in June/July 2022 (I mean it this time!).  Keep the requests coming, sorry for the delays!

Thanks!
Amanda

**Comment by sta-c0000** [comment]
Just started playing with OrigamiSimulator and it's great, thank you!

My favourite existing additional feature so far is export to OBJ and that it includes UV map!  This allows us to not only texture but also projection map images onto folded 3D paper back into the unfolded texturemap, which can then be printed.  Fantastic!

As for additional features, I really appreciate the face/push/pull functionality of [webOAcard](https://github.com/caa1211/webOAcard), but not sure yet how to interface/exchange data with OrigamiSimulator.  Perhaps something would need to be done on the webOAcard side.  But, yes, in terms of editing, that face/push/pull feature for kirigami popups is great.
Thanks!

**Reply by amandaghassaei** [reply]
thanks for that link, was not aware of webOAcard.  Yeah I think something like a push pull could be possible in 3D, at least up to certain limits.  Thanks for the suggestion!

**Comment by forresto** [comment]
> Separate the simulation and other mathy stuff into a core library that can be imported into other projects (via npm) and runs headless.

I like this idea. Another angle would be an "embeddable" version of the app that could take `svg, options` as input, and display the solved output. If you could embed this in a webpage in an iframe, and control it via JS, that would open a few possibilities.

* Origami artists could embed into their own sites
* Before there are animation tools build into Origami Simulator, somebody could script animating parameters, and capture the iframe
* You could have a procedural design generator that feeds into the solver, previewed in real-time.

I've been working on https://cuttle.xyz and we've been discussing a similar embeddable player for our published projects. If you could have both projects running in the same page, you could change parameter in the Cuttle design, and see the Origami Simulator update in real time. Here's Jun Mitani's [2010126 Origami Sphere (16 flaps)](https://mitani.cs.tsukuba.ac.jp/en/cp_download.html) made in Cuttle, with a parameter that tweaks the curve. Instead of downloading the SVG from Cuttle and uploading it to Origami Simulator, there could be a one-page webapp that plugs them together, for more fluid feedback.

https://user-images.githubusercontent.com/395307/113452341-777aff80-940c-11eb-89cd-8531b48e6072.mov

So in a nutshell, the feature request is for a scriptable, embeddable "player."

**Reply by amandaghassaei** [reply]
Nice!  Hadn't considered something like this, but should be really easy to do once the core simulation stuff has been consolidated into a separate library.  Cool app btw - would be fantastic to get Cuttle working with the sim - would save me a lot of work for sure building out a 2d interface.  I think the scriptability of your tool would be very helpful for origami design, we should chat more for sure!

**Reply by forresto** [reply]
A smaller solution to make interop smoother now: accept svg string paste: https://github.com/amandaghassaei/OrigamiSimulator/pull/49

**Comment by smmsamm** [comment]
I really enjoy your beautiful program,
Would you please add a feature for putting custom jpg/png as texture?
Or maybe App has this abilities now and I don't see?
thank you

**Reply by edemaine** [reply]
This is issue #13. I agree it would be useful!

**Comment by siliffe** [comment]
Hi there, I think it would be a great to have a 'glue' feature, where you can select sections of the folded model that would not unfold, Is this possible? some images attached to show the difference in unfolding if a line on the model was glued (or stitched in this case
![IMG_1097](https://user-images.githubusercontent.com/41992163/145033158-791d2e4c-9e70-406c-9fb6-d82ae36c4cb2.jpeg)
![IMG_1099-2](https://user-images.githubusercontent.com/41992163/145033164-e550fe99-8290-4a38-95db-fe4db02d9e1b.jpeg)
![IMG_1100-3](https://user-images.githubusercontent.com/41992163/145033168-1434a783-8903-4563-8092-72b23644e9b8.jpeg)
![IMG_1101-4](https://user-images.githubusercontent.com/41992163/145033173-18aed36f-9d58-4ffb-81de-676d2ab1c709.jpeg)
)

**Reply by amandaghassaei** [reply]
thanks for the request!  This is interesting, I was imagining that we'd add an edge/edge glue functionality, but this is a case where the gluing is happening from an edge to the middle of a face or even between faces in a certain region.  Maybe to simplify the problem to start we could require that this line is drawn on the crease pattern (in some non-folding color), and then when it's folded up it can be selected for the glue operation (same as an edge/edge glue).  Would that work?

**Reply by siliffe** [reply]
Yes, that could work and I think it would be fairly straightforward to find out which bits need to be glued on the creasepattern. I also thought maybe you could morph the function that allows some creases to fold after others (which i think is already there?) to just never unfold, and then choose small sections of the crease lines to be this type of crease

**Reply by gkonjevod** [reply]
How about gluing nodes to nodes? That would let one glue edges to edges or to points elsewhere (as long as a node exists). For my situation, I'd be happy with a way to specify this through .fold files.

**Reply by gkonjevod** [reply]
Oh, and another twist on the gluing would be to allow one to fix a node or several in space, the idea being that we could simulate a folded piece partially unfolding and hanging from one or more suspension points due to its own weight.

**Reply by edemaine** [reply]
> For my situation, I'd be happy with a way to specify this through .fold files.

You should already be able to do something like this by merging the relevant nodes.  To glue nodes A and B, just replace all instances of node B with node A (and optionally delete node B).

But this will only get the right lengths if A and B were collocated to begin with.  So if you're folding a cylinder, you'd need to manually (in software) wrap the FOLD file onto a cylinder to make the two edges meet, then combine the vertices on those edges.

A nice small feature would be if Origami Simulator respected `edges_length` as an override for desired edge lengths.  Then it would be enough to set this according to the 2D geometry, and you can fuse vertices to your heart's content. This feature shouldn't be too hard to add, but it's annoying that you'd still need to compute the lengths manually (and probably triangulate first, so that the lengths determine the intrinsic geometry).

I should mention that [Freeform Origami](https://origami.c.u-tokyo.ac.jp/~tachi/software/) has (I believe) the ability to interactively weld meshes, which does something like what I'm describing (and preserving the intrinsic geometry).  It should also save as FOLD, so you could bring the resulting mesh back into Origami Simulator if you wanted.

**Reply by gkonjevod** [reply]
Oh, that should work! I can generate the locations of the nodes I'd like to glue easily in the flat-folded state. I'll try it out--thank you!

**Reply by amandaghassaei** [reply]
The FOLD import has not been stress tested as much as I'd like, so I'd be interested to hear if bringing in non-manifold vertices works as intended.

**Reply by siliffe** [reply]
> You should already be able to do something like this by merging the relevant nodes. To glue nodes A and B, just replace all instances of node B with node A (and optionally delete node B).
> 

How would this actually be done in the .fold file? I see vertices_coords, edges_vertices, edges_assignment and edges_foldAngle but not 'node'.

**Reply by amandaghassaei** [reply]
fields like edges_vertices and faces_vertices refer to vertices by index, so you need to replace all instances of vertex A with vertex B in those fields by swapping one index for another.  Deleting a vertex would mean removing it from vertices_coords – but you have to be careful to update all the relevant indices in the other fields in case the deletion caused the other vertex indices to change.  There may be methods in the FOLD repo to help with some of these operations: https://github.com/edemaine/fold/blob/main/doc/api.md
if not, we should add them.

**Reply by siliffe** [reply]
hi there, so after learning a bit more about the .fold file I tried to glue a pleat together by changing some of the co-ordinates of the vertices and edges. in the image from my notebook I have tried to make v3 into v6 and v4 into v8. However when I tried this in the origami simulator I got the images shown below (the paper is no longer square). I have also attached the different methods I tried in the code of 'gluing', v1 was just changing the vertices_coords, v3 was changing vertices_edges as well
<img width="903" alt="Screenshot 2021-12-22 at 17 36 39" src="https://user-images.githubusercontent.com/41992163/147136026-70cf2044-d4a4-44e1-bfe5-297f3db743d2.png">
<img width="903" alt="Screenshot 2021-12-22 at 17 36 27" src="https://user-images.githubusercontent.com/41992163/147136046-1dddeeb1-212b-4f3e-a458-8ce8c247cdb7.png">
<img width="903" alt="Screenshot 2021-12-22 at 17 35 51" src="https://user-images.githubusercontent.com/41992163/147136064-431e7dae-2d62-469f-9167-00214dec597e.png">
![IMG_1177](https://user-images.githubusercontent.com/41992163/147136087-7e6b6b88-a150-42e1-8af5-c72a60fa549e.jpg)
[FoldFiles.zip](https://github.com/amandaghassaei/OrigamiSimulator/files/7764423/FoldFiles.zip)

**Comment by evbernardes** [comment]
> Ability to connect edges in a design so that you can import a tessellation as SVG and turn it into e.g. a connected tube in 3D. Currently this type of thing is only supported through fold, but not everyone wants to write fold files.

This feature would be absolutely amazing! I've been coding an Inkscape tesselation plugin for fun, and it would be awesome to be able to just use the exact same svg file both for simulation and for laser cutting!

**Comment by gkonjevod** [comment]
Hi Amanda,
I would love to have ways to allow some creases to fold differently than others--either at a different rate (for example, I might know that in the final model, I want some creases to be only half-folded), or to specify the sequence in which the folds are activated. I'm not too concerned about the UI, because I usually generate .fold files.

**Reply by edemaine** [reply]
Hi Goran, you can already specify target folding amounts with `edges_foldAngle` in FOLD format, or with opacity in SVG. Does this capture what you mean by "rates"?

I'm not sure how you'd specify sequencing, or even what it would mean... Perhaps a sequence of frames, and each frame has its own `edges_foldAngle`?

**Reply by gkonjevod** [reply]
Hi Erik,
Thanks for the response! I'll check out the foldAngle feature.
For sequencing, I'm thinking of something like my pleat tessellations where I could specify individual pleats and simulate them getting folded one by one. Are frames already in FOLD? (Sorry for asking; I should just check that myself.)

**Reply by edemaine** [reply]
Frames are part of the FOLD spec, but I don't think they're supported in any software yet. So I think supporting some form of frames in Origami Simulator is a possible/reasonable interpretation of your "sequence" request.

**Reply by gkonjevod** [reply]
OK, I've tried an example with edges_foldAngle and it seems to work. Thanks!

**Reply by amandaghassaei** [reply]
FOLD frames are not yet supported, but I'll add that to the list, thanks Goran!

**Comment by gkonjevod** [comment]
Here's a fold file that gets launched into space by the simulator... I'm sure there are bugs in the code I'm using to merge the vertices, but still, it's a behavior I haven't seen before.
[simple_tess_1_2_merged.zip](https://github.com/amandaghassaei/OrigamiSimulator/files/7738987/simple_tess_1_2_merged.zip)

**Reply by edemaine** [reply]
This might be worth a separate discussion/issue, and I don't know what's going on, but for debugging it can help to Pause Simulation, Re-center geometry, and Step Forward.  You can at least see what's going on, though I don't know whether this helps:

![image](https://user-images.githubusercontent.com/2218736/146650067-3832899c-0566-4d9a-9f59-c5eb71425509.png)

**Reply by gkonjevod** [reply]
Thank you (I forgot about those features), but it doesn't really help :) I don't think it's what I expect to see, so I'll have to go through the nodes by hand to see what I actually merged. The simpler instance (one pair of pleats) seems to work, but in that one the "supervertices" form a single connected component and here they don't.

**Comment by david-dufour-kw** [comment]
It would be interesting if I could provide a 3D volume and an algorithm would optimize to fill the shape as much as possible. A form of AI assisted origami creation.

**Reply by gkonjevod** [reply]
Check out Tomohiro Tachi's Origamizer [https://origami.c.u-tokyo.ac.jp/~tachi/software/](url)

**Comment by amandaghassaei** [comment]
some of you have already found this, but feel free to fill out this short user feedback survey: https://docs.google.com/forms/d/e/1FAIpQLSeWp5yQkzONw9YjhrkZG-9NJCisgXCk0ma23-yoitI7PBykFQ/viewform?usp=send_form

I know I've been promising some updates to this app for several years (sorry!), but I'm actually working on it now!

**Comment by MatthewElveyPrice** [comment]
I'm currently planning out tools for curved origami, I think my plan is similar to your description of "a dynamic triangulation that discretizes the surface more where bending is detected.". Specifically, I want to try triangulation that isn't purely "more detail", but aligning the lines between polygons to the curvature of the paper.

I expect it will take me a while to produce anything - checking in here to avoid rework and so on, in case some of the points are already in progress somewhere, including other bits like switching to typescript / webpack, refactoring libraries or switching UI
